Hazel/src/Hazel/ImGui/ImGuiConsole.cpp

```diff
@@ -0,0 +1,249 @@
+#include "hzpch.h"
+#include "ImGuiConsole.h"
+
+#include "imgui.h"
+#include "examples/imgui_impl_glfw.h"
+#include "examples/imgui_impl_opengl3.h"
+
+namespace Hazel {
+
+	float ImGuiConsole::s_DisplayScale = 1.0f;
+	uint16_t ImGuiConsole::s_MessageBufferCapacity = 200;
+	uint16_t ImGuiConsole::s_MessageBufferSize = 0;
+	uint16_t ImGuiConsole::s_MessageBufferBegin = 0;
+	ImGuiConsole::Message::Level ImGuiConsole::s_MessageBufferRenderFilter = ImGuiConsole::Message::Level::Trace;
+	std::vector<Ref<ImGuiConsole::Message>> ImGuiConsole::s_MessageBuffer(s_MessageBufferCapacity);
+	bool ImGuiConsole::s_AllowScrollingToBottom = true;
+	bool ImGuiConsole::s_RequestScrollToBottom = false;
+
+	void ImGuiConsole::AddMessage(Ref<Message> message)
+	{
+		if (message->m_Level == Message::Level::Invalid)
+			return;
+
+		*(s_MessageBuffer.begin() + s_MessageBufferBegin) = message;
+		if (++s_MessageBufferBegin == s_MessageBufferCapacity)
+			s_MessageBufferBegin = 0;
+		if (s_MessageBufferSize < s_MessageBufferCapacity)
+			s_MessageBufferSize++;
+
+		// Request to scroll to bottom of the list to view the new message
+		s_RequestScrollToBottom = s_AllowScrollingToBottom;
+	}
+
+	void ImGuiConsole::Clear()
+	{
+		for (auto message = s_MessageBuffer.begin(); message != s_MessageBuffer.end(); message++)
+			(*message) = std::make_shared<Message>();
+		s_MessageBufferBegin = 0;
+	}
+
+	void ImGuiConsole::OnImGuiRender(bool* show)
+	{
+		ImGui::SetNextWindowSize(ImVec2(640, 480), ImGuiCond_FirstUseEver);
+		ImGui::Begin("Console", show);
+		{
+			ImGuiRendering::ImGuiRenderHeader();
```

Comment:
One thing to be careful when naming things similar to the API for ImGui is that a reader could misunderstand and think this is an ImGui API when instead its part of the console logging API